### ISSUE TITLE
Changes needed in javascript file glymagesvg.js for glycan highlighting widget.

### ISSUE THREAD

**Issue by sujeetvkulkarni** [glygener/glygen-issues#855]
Please find below the proposed changes,

1. Add a top level function in glymagesvg.js (not part of glymagesvg variable) which takes params - partial json url and partial svg url and calls glymagesvg.init() internally passing the params. Ignore baseurl param/getglymagesvgloc() for url construction.
2. Use accession from div and above params to create json, svg urls to download, based on a predefined strategy. 
3. Remove logic to add css file - or at least make it configurable.
 ```
       var head = document.getElementsByTagName('head')[0];
        if (!(head.innerHTML.includes('glymagesvg.css'))) {
            head.innerHTML += '<link rel="stylesheet" href="'+glymagesvg.params.baseurl+'/css/glymagesvg.css" type="text/css" />';
        }
```
5. Below line contains svg_file variable which is not declared or assigned any value,			
```  
          console.log("I need an svg file:", svg_file, response.text())
```
6. Adding  glymagesvg_annotation={"MotifAlignments." + id} attribute in div highlights motif and sets opacity for glycan image but initiating click event on corresponding <span> does not set opacity back to normal. You have click twice. One click should set opacity to normal.

```
             <div id="glymagesvg3"
                glymagesvg_accession={glytoucan_ac}
                glymagesvg_annotation={"MotifAlignments." + id}
                glymagesvg_imageclass="glymagesvg_low_opacity"
                glymagesvg_monoclass="glymagesvg_high_opacity"
                glymagesvg_linkclass="glymagesvg_high_opacity"
                glymagesvg_linkinfoclass="glymagesvg_high_opacity"
                glymagesvg_monoclick_highlights_parent_link="true"
                glymagesvg_monoclick_highlights_related_monos="true"
            />
            <span glymagesvg_forid="glymagesvg3" glymagesvg_annotation = {"MotifAlignments." + id} >motif
            </span>
```

**Comment by edwardsnj** [comment]
Item 1, 2: See https://glymage.glyomics.org/demo/demo2.html for a method for arbitrary SVG/JSON urls. This is sufficient. 

Item 3. How do you actually want to specify the CSS file? Do you want the glymage one or one or your own? 

Item 5, I will delete this line, doesn't do anything. Note that I fixed the error handling since the last time you picked up the JS file, so you should get the version I committed last week.

Item 6, you are listing the glymagesvg_annotation attribute on *both* the figure /and/ the clickable items. If you are doing remote clicks, glymagesvg_annotation should be on the clickable items only. I'm not sure if that is the reason for the behavior you are experiencing.

**Comment by edwardsnj** [comment]
Note that the monoclick attributes have changed in the new version.

https://edwardslab.bmcb.georgetown.edu/glymagedev/demo/demo2.html

shows a CSS parameter option - cssurl. You can also set this attribute to null to avoid automatic css and do it yourself. 

With respect to item 6, it might be you are hitting deliberate behavior. Once you have removed the extra annotation attribute, we can discuss. For some interpretations of what you are asking for, the current behavior is deliberate - clicking is one direction at a time (figure -> remotes, or remotes -> figure) when you switch to the other click target, it removes any state and then applies the positive click. The behavior you might be asking for only makes sense for very limited classes of annotations (and for a very limited click sequences). Happy to discuss at length. I spent a long time playing with different options to make sure it was powerful, flexible, and predictable for the user.

**Comment by edwardsnj** [comment]
@sujeetvkulkarni @ReneRanzinger I've prototyped up a version of what I understood you wanted for the motif page. See what you think. There are still some bugs I'm figuring out, but I think it basically works: https://edwardslab.bmcb.georgetown.edu/glymagedev/demo/demo4.html

**Comment by sujeetvkulkarni** [comment]
@edwardsnj thanks for making the changes. I have checked the demo page, looks like it works. Please let us know when changes are done, so that I can update the javascript file at our end to test it. I have to specifically check how it works in pagination case.

**Comment by edwardsnj** [comment]
There is still a bug in there I need to figure out. I will check in the current version, but there is something still not quite right about the highlight_parent_link option. 

For paging/dynamically generating elements with glymagesvg_* attributes in them, here is the code I'm using in glycomotif: 

https://github.com/glygen-glycan-data/PyGly/blob/1855f0b6584b6828cfa83e72cbd503fc23c9a27e/smw/glycomotif/wiki/mediawiki/Common.js.txt#L512

and then

https://github.com/glygen-glycan-data/PyGly/blob/1855f0b6584b6828cfa83e72cbd503fc23c9a27e/smw/glycomotif/wiki/mediawiki/Common.js.txt#L609

Basically create your HTML elements, add the glymagesvg attributes and then call glymagesvg.init again. glymagesvg.init knows whether it has processed an HTML element before, so it doesn't add another SVG image to containers it has already put one in.

**Comment by edwardsnj** [comment]
Bug with javascript functions for turning highlighting on/off fixed. New javascript file checked in. There are some outstanding issues with how the link for undefined toplogies are highlighted, and sulfated HexN residues. But this will resolve in time.

**Comment by sujeetvkulkarni** [comment]
@edwardsnj thanks for the update and sharing GlycoMotif links. I will use the latest javascript file and incorporate the changes.

**Comment by edwardsnj** [comment]
@sujeetvkulkarni latest version of glymage.js adds an option to control whether clicks are handled in various places:
`    glymagesvg_clicktarget="remote_element,figure_mono,figure_background"
`
Remove one or more of these comma-separated terms to disable click handling.

You probably want: 
`    glymagesvg_clicktarget="remote_element,figure_background"
`
or
`    glymagesvg_clicktarget="remote_element"
`
depending on whether you want to allow users to click the background of the image to reset the highlight state. 

See:  https://edwardslab.bmcb.georgetown.edu/glymagedev/demo/demo3.html

**Comment by sujeetvkulkarni** [comment]
@edwardsnj Thanks a lot, I will take the latest file and work on integrating these changes in GlyGen.

**Comment by edwardsnj** [comment]
implemented in GlyGen, no outstanding current issues
